File: leetcode/3_longest_substring_without_repeat_char/3_ans.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
    	HashMap<Character, Integer> map = new HashMap<Character, Integer>();
    	int start=0;
    	int lengthValue=0;
    	
    	for(int i=0;i<s.length();i++){
    		
    		
    		
    		if(map.containsKey(s.charAt(i))){	
    			
    			char tempC = s.charAt(i);
    			
    			if(start<map.get(s.charAt(i))+1){
    				start = map.get(s.charAt(i)) + 1;
    			}
    			//map.clear();
    			map.put(tempC, i);
    		}else{
    			map.put(s.charAt(i), i);	// char_index map
    		}
    		//System.out.println(i+"\t"+start+"\t"+(i-start+1));	
    		
    		if((i-start+1 > lengthValue) )
    			lengthValue=i-start+1;	
    		
    		
    		
    	}
    	if(s.length()==1)
    		lengthValue=1;
    	
    	//System.out.println(lengthValue);
    	return lengthValue;
    }
}

File: leetcode/3_longest_substring_without_repeat_char/3_ans.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        lookup = {}
        start = 0 
        length = 0
        
        for i in range(len(s)):
            
            if s[i] in lookup:
                
           
                if  start < lookup[s[i]]+1:
                    start = lookup[s[i]]+1
                lookup[s[i]] = i
                
            else:
                lookup[s[i]] = i
            
            if i-start+1 > length:
                length = i-start+1
            
            if len(s)==1:
                length = 1
          
            
        return length
